In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('Cleaned Data with States.csv')
df.head()

Unnamed: 0  Alcohol                                   Appellation Category  \
0           0     14.9  Russian River Valley, Sonoma, California, US      Red   
1           1     14.5  Central Coast, Central Coast, California, US      Red   
2           2     13.8                           California Republic      Red   
3           3     13.8          Lodi, Central Valley, California, US      Red   
4           4     13.0                                New Jersey, US      Red   

   Points  Price                                             Review  \
0      87  42.00  Baked plum, licorice and lavender aromas and f...   
1      87  19.98  Pomegranate and light baking-spice aromas show...   
2      86  15.00  Butter and vanilla notes dominate the jammy fr...   
3      86  10.00  An aroma like toasted almonds and wood smoke m...   
4      86  27.00  Plum skin and pomegranate lead the nose while ...   

      Variety                                               Wine  \
0   Zinfandel  V. Sattui 2015 Gilsson Vineyard Old Vine Zinfa...   
1  Pinot Noir  Wente 2014 Coastal Selection Pinot Noir (Centr...   
2  California  California Republic 2016 Cabernet Sauvignon (C...   
3  Pinot Noir     Collier Creek 2015 Red Wagon Pinot Noir (Lodi)   
4   Red Blend                  DiLuca 2016 Rosso Black Label Red   

                Winery                  Town         Broad_Region  \
0            V. Sattui  Russian River Valley               Sonoma   
1                Wente         Central Coast        Central Coast   
2  California Republic   California Republic  California Republic   
3        Collier Creek                  Lodi       Central Valley   
4               DiLuca            New Jersey                   US   

                 State Broad_Region.1    Year                 Var  
0           California            NaN  2015.0           Zinfandel  
1           California            NaN  2014.0          Pinot Noir  
2  California Republic     California  2016.0  Cabernet Sauvignon  
3           California            NaN  2015.0          Pinot Noir  
4                   US            NaN  2016.0           Red Blend

In [3]:
df['Broad_Region.1'].value_counts(normalize=False, sort=True, ascending=True, bins=None, dropna=False)

Yorkville Highlands                 1
Ribbon Ridge                        1
Seneca Lake                         1
Dunnigan Hills                      1
Long Island                         1
Texas                               1
Marsanne-Roussanne                  1
Alta Mesa                           1
New York                            1
Rockpile                            1
Marsanne-Viognier                   1
Chalk Hill                          1
Texas High Plains                   1
Mendocino Ridge                     1
Yadkin Valley                       1
McMinnville                         1
Paso Robles Estrella District       1
Mt. Harlan                          1
Sonoma Mountain                     1
Ancient Lakes                       1
Mission                             1
Arroyo Grande Valley                1
Norton                              1
Touriga Nacional                    1
Paicines                            1
South Coast                         1
Maryland    

In [4]:
df = df.drop(columns=['Unnamed: 0','Broad_Region.1'])
df.head()

Alcohol                                   Appellation Category  Points  \
0     14.9  Russian River Valley, Sonoma, California, US      Red      87   
1     14.5  Central Coast, Central Coast, California, US      Red      87   
2     13.8                           California Republic      Red      86   
3     13.8          Lodi, Central Valley, California, US      Red      86   
4     13.0                                New Jersey, US      Red      86   

   Price                                             Review     Variety  \
0  42.00  Baked plum, licorice and lavender aromas and f...   Zinfandel   
1  19.98  Pomegranate and light baking-spice aromas show...  Pinot Noir   
2  15.00  Butter and vanilla notes dominate the jammy fr...  California   
3  10.00  An aroma like toasted almonds and wood smoke m...  Pinot Noir   
4  27.00  Plum skin and pomegranate lead the nose while ...   Red Blend   

                                                Wine               Winery  \
0  V. Sattui 2015 Gilsson Vineyard Old Vine Zinfa...            V. Sattui   
1  Wente 2014 Coastal Selection Pinot Noir (Centr...                Wente   
2  California Republic 2016 Cabernet Sauvignon (C...  California Republic   
3     Collier Creek 2015 Red Wagon Pinot Noir (Lodi)        Collier Creek   
4                  DiLuca 2016 Rosso Black Label Red               DiLuca   

                   Town         Broad_Region                State    Year  \
0  Russian River Valley               Sonoma           California  2015.0   
1         Central Coast        Central Coast           California  2014.0   
2   California Republic  California Republic  California Republic  2016.0   
3                  Lodi       Central Valley           California  2015.0   
4            New Jersey                   US                   US  2016.0   

                  Var  
0           Zinfandel  
1          Pinot Noir  
2  Cabernet Sauvignon  
3          Pinot Noir  
4           Red Blend

In [5]:
df.describe(include = 'all')

Alcohol                                   Appellation Category  \
count   4768.000000                                          4768     4768   
unique          NaN                                           945        7   
top             NaN  Russian River Valley, Sonoma, California, US      Red   
freq            NaN                                           214     3118   
mean      13.949778                                           NaN      NaN   
std        0.950407                                           NaN      NaN   
min        8.000000                                           NaN      NaN   
25%       13.500000                                           NaN      NaN   
50%       14.100000                                           NaN      NaN   
75%       14.500000                                           NaN      NaN   
max       22.000000                                           NaN      NaN   

             Points        Price  \
count   4768.000000  4768.000000   
unique          NaN          NaN   
top             NaN          NaN   
freq            NaN          NaN   
mean      89.340185    42.290271   
std        2.784258    32.311230   
min       80.000000     5.000000   
25%       87.000000    24.000000   
50%       90.000000    35.000000   
75%       91.000000    50.000000   
max      100.000000   475.000000   

                                                   Review     Variety  \
count                                                4768        4768   
unique                                               4678         228   
top     Green aromas like pine needles and celery are ...  Pinot Noir   
freq                                                    2         862   
mean                                                  NaN         NaN   
std                                                   NaN         NaN   
min                                                   NaN         NaN   
25%                                                   NaN         NaN   
50%                                                   NaN         NaN   
75%                                                   NaN         NaN   
max                                                   NaN         NaN   

                                               Wine    Winery  \
count                                          4768      4768   
unique                                         4676      1520   
top     Naggiar 2017 Estate Rosé (Sierra Foothills)  Maryhill   
freq                                              3        38   
mean                                            NaN       NaN   
std                                             NaN       NaN   
min                                             NaN       NaN   
25%                                             NaN       NaN   
50%                                             NaN       NaN   
75%                                             NaN       NaN   
max                                             NaN       NaN   

                        Town    Broad_Region        State         Year  \
count                   4768            4768         4768  4768.000000   
unique                   943             800          773          NaN   
top     Russian River Valley   Central Coast   California          NaN   
freq                     214             780         2207          NaN   
mean                     NaN             NaN          NaN  2015.419044   
std                      NaN             NaN          NaN     1.154587   
min                      NaN             NaN          NaN  2004.000000   
25%                      NaN             NaN          NaN  2015.000000   
50%                      NaN             NaN          NaN  2015.000000   
75%                      NaN             NaN          NaN  2016.000000   
max                      NaN             NaN          NaN  2017.000000   

               Var  
count         4768  
unique         117  
top     Pinot Noir  
freq       

In [6]:
len(df)

4768

In [7]:
df = df.dropna()
len(df)

4768

In [8]:
#dropping NLTK vars
df = df.drop(columns=['Review','Variety','Wine','Winery','Town','Broad_Region','State'])
df.head()

Alcohol                                   Appellation Category  Points  \
0     14.9  Russian River Valley, Sonoma, California, US      Red      87   
1     14.5  Central Coast, Central Coast, California, US      Red      87   
2     13.8                           California Republic      Red      86   
3     13.8          Lodi, Central Valley, California, US      Red      86   
4     13.0                                New Jersey, US      Red      86   

   Price    Year                 Var  
0  42.00  2015.0           Zinfandel  
1  19.98  2014.0          Pinot Noir  
2  15.00  2016.0  Cabernet Sauvignon  
3  10.00  2015.0          Pinot Noir  
4  27.00  2016.0           Red Blend

## For Scaling:
- Alcohol
- Points
- Price

## for dummy variables:
- appellation (945 unique)
- category (7 unique)
- var (117 unique)
- year 

## NLTK for:
- Reviews
- Wine
- Winery


In [9]:
df = pd.get_dummies(df, columns=['Year','Var','Appellation'],drop_first = True)
df.head()

Alcohol Category  Points  Price  Year_2005.0  Year_2007.0  Year_2008.0  \
0     14.9      Red      87  42.00            0            0            0   
1     14.5      Red      87  19.98            0            0            0   
2     13.8      Red      86  15.00            0            0            0   
3     13.8      Red      86  10.00            0            0            0   
4     13.0      Red      86  27.00            0            0            0   

   Year_2009.0  Year_2010.0  Year_2011.0         ...          \
0            0            0            0         ...           
1            0            0            0         ...           
2            0            0            0         ...           
3            0            0            0         ...           
4            0            0            0         ...           

   Appellation_Yakima Valley, Columbia Valley, Washington, US  \
0                                                  0            
1                                                  0            
2                                                  0            
3                                                  0            
4                                                  0            

   Appellation_Yamhill-Carlton, Willamette Valley, Oregon, US  \
0                                                  0            
1                                                  0            
2                                                  0            
3                                                  0            
4                                                  0            

   Appellation_Yorkville Highlands, North Coast, California, US  \
0                                                  0              
1                                                  0              
2                                                  0              
3                                                  0              
4                                                  0              

   Appellation_Yountville, Napa, California, US  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   

   Appellation_Z. Alexander Brown  Appellation_ZD  Appellation_Ziata  \
0                               0               0                  0   
1                               0               0                  0   
2                               0               0                  0   
3                               0               0                  0   
4                               0               0                  0   

   Appellation_Zinfandelic  Appellation_Zotovich  Appellation_Écluse  
0                        0                     0                   0  
1                        0                     0                   0  
2                        0                     0                   0  
3                        0                     0                   0  
4                        0                     0                   0  

[5 rows x 1076 columns]

## Feature engineering - target = Category/Color

In [10]:
features = df.drop(columns=['Category'])
target = df['Category']

In [11]:
target.value_counts(normalize=False, sort=True, ascending=True, bins=None, dropna=True)

Port/Sherry       4
Fortified         4
Dessert          15
Sparkling        70
Rose            208
White          1349
Red            3118
Name: Category, dtype: int64

In [12]:
features.head()

Alcohol  Points  Price  Year_2005.0  Year_2007.0  Year_2008.0  Year_2009.0  \
0     14.9      87  42.00            0            0            0            0   
1     14.5      87  19.98            0            0            0            0   
2     13.8      86  15.00            0            0            0            0   
3     13.8      86  10.00            0            0            0            0   
4     13.0      86  27.00            0            0            0            0   

   Year_2010.0  Year_2011.0  Year_2012.0         ...          \
0            0            0            0         ...           
1            0            0            0         ...           
2            0            0            0         ...           
3            0            0            0         ...           
4            0            0            0         ...           

   Appellation_Yakima Valley, Columbia Valley, Washington, US  \
0                                                  0            
1                                                  0            
2                                                  0            
3                                                  0            
4                                                  0            

   Appellation_Yamhill-Carlton, Willamette Valley, Oregon, US  \
0                                                  0            
1                                                  0            
2                                                  0            
3                                                  0            
4                                                  0            

   Appellation_Yorkville Highlands, North Coast, California, US  \
0                                                  0              
1                                                  0              
2                                                  0              
3                                                  0              
4                                                  0              

   Appellation_Yountville, Napa, California, US  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   

   Appellation_Z. Alexander Brown  Appellation_ZD  Appellation_Ziata  \
0                               0               0                  0   
1                               0               0                  0   
2                               0               0                  0   
3                               0               0                  0   
4                               0               0                  0   

   Appellation_Zinfandelic  Appellation_Zotovich  Appellation_Écluse  
0                        0                     0                   0  
1                        0                     0                   0  
2                        0                     0                   0  
3                        0                     0                   0  
4                        0                     0                   0  

[5 rows x 1075 columns]

In [13]:
import numpy as np
from matplotlib import pyplot as plt
from sklearn import datasets, linear_model
from sklearn.model_selection import train_test_split
from sklearn import metrics
X_train, X_test, y_train, y_test = train_test_split(features, target, random_state=32,test_size=0.2)

In [14]:
from sklearn import preprocessing
from sklearn import pipeline

scaler = preprocessing.StandardScaler()
scaler

StandardScaler(copy=True, with_mean=True, with_std=True)

In [15]:
scaler.fit(features)

/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)


StandardScaler(copy=True, with_mean=True, with_std=True)

In [16]:
#need to scale the non dummy vars
X_train.iloc[:,0:3].head()

Alcohol  Points  Price
1657     14.2      89   30.0
2559     13.7      83   22.0
4253     13.7      84   22.0
2139     14.4      88   23.0
2897     14.5      92   60.0

In [17]:
scaler.fit(X_train.iloc[:,0:3])
features_scaled_train = pd.DataFrame(scaler.transform(X_train.iloc[:,0:3]), columns=X_train.columns[:3], index=X_train.index)


/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  


In [18]:
list(features.columns[3:-1])[0:10]

['Year_2005.0',
 'Year_2007.0',
 'Year_2008.0',
 'Year_2009.0',
 'Year_2010.0',
 'Year_2011.0',
 'Year_2012.0',
 'Year_2013.0',
 'Year_2014.0',
 'Year_2015.0']

In [19]:
non_scaled_featues = list(features.columns[3:-1])
for feature in non_scaled_featues:
    features_scaled_train[feature]=X_train[feature]

In [20]:
features_scaled_train

Alcohol    Points     Price  Year_2005.0  Year_2007.0  Year_2008.0  \
1657  0.279669 -0.116308 -0.389654            0            0            0   
2559 -0.247591 -2.267960 -0.648111            0            0            0   
4253 -0.247591 -1.909352 -0.648111            0            0            0   
2139  0.490573 -0.474917 -0.615804            0            0            0   
2897  0.596025  0.959518  0.579557            0            0            0   
2195 -0.353043 -0.474917 -0.066584            0            0            0   
4023  1.966900 -0.474917 -0.325040            0            0            0   
1708 -0.985754 -0.474917 -0.680418            0            0            0   
1203 -1.091206 -0.833526 -0.648111            0            0            0   
1776 -0.194865 -0.474917 -0.551190            0            0            0   
444   0.701477  0.242301  1.871840            0            0            0   
1950  1.017832 -1.192134 -0.098891            0            0            0   
3411  0.279669 -0.833526 -0.551190            0            0            0   
3533 -0.142139 -1.192134 -0.551190            0            0            0   
2224  0.806928 -0.474917  0.062644            0            0            0   
1202  0.912380 -0.833526 -0.131198            0            0            0   
1279 -0.985754 -1.550743 -0.906567            0            0            0   
3095 -0.036687  0.242301  0.418022            0            0            0   
2663  0.596025  0.959518  1.387234            0            0            0   
838   0.806928 -1.192134 -0.131198            0            0            0   
1967  1.334188 -1.550743 -0.001970            0            0            0   
4199 -0.036687 -1.550743 -0.228119            0            0            0   
3509 -0.142139 -1.192134 -0.809646            0            0            0   
580   0.806928 -0.116308 -0.131198            0            0            0   
3139  0.174217  0.242301  0.579557            0            0            0   
3542  0.490573 -1.192134  0.224180            0            0            0   
90   -0.669399 -1.909352  0.094951            0            0            0   
4336  0.490573  1.318127  0.418022            0            0            0   
729   0.490573 -0.833526 -0.066584            0            0            0   
902  -0.036687  1.318127  3.487193            0            0            0   
...        ...       ...       ...          ...          ...          ...   
1114  1.397459  0.242301 -0.389654            0            0            0   
4660 -0.669399  0.242301  0.385715            0            0            0   
2522  0.596025 -1.550743  0.224180            0            0            0   
2168 -0.142139 -0.474917 -0.551190            0            0            0   
4491 -0.142139  0.600909  0.579557            0            0            0   
3661  0.596025  1.318127 -0.066584            0            0            0   
1190 -0.321407 -0.833526 -0.389654            0            0            0   
4345  0.596025  1.318127  1.225699            0            0            0   
185  -0.458495  1.318127 -0.777339            0            0            0   
100  -0.985754  2.035344  0.579557            0            0            0   
3506  0.806928 -1.192134 -0.777339            0            0            0   
2004 -0.690489 -1.550743 -0.551190            0            0            0   
650   0.596025 -0.474917 -0.551190            0            0            0   
1578  0.596025  0.242301  0.288794            0            0            0   
35   -0.985754 -1.192134 -0.389654            0            0            0   
1442  0.596025  0.959518 -0.648111            0            0            0   
2625  0.068765  1.318127  1.548769            0            0            0   
1283  0.174217 -1.550743  0.094951            0            0            0   
9     0.279669 -1.192134  0.094951            0            0            0   
1258  0.596025 -1.192134 -0.874260            0            0            0   
384

In [21]:
scaler.fit(X_test.iloc[:,0:3])
features_scaled_test = pd.DataFrame(scaler.transform(X_test.iloc[:,0:3]), columns=X_test.columns[:3], index=X_test.index)


/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  


In [22]:
# non_scaled_featues = list(features.columns[3:-1])
for feature in non_scaled_featues:
    features_scaled_test[feature]=X_test[feature]

In [23]:
features_scaled_test

Alcohol    Points     Price  Year_2005.0  Year_2007.0  Year_2008.0  \
169  -1.579266  1.300871  0.075038            0            0            0   
2092 -0.637792  1.300871  0.155638            0            0            0   
2607 -0.323967  1.662582  0.773569            0            0            0   
4623 -0.533184  0.215738 -0.811558            0            0            0   
429  -0.533184  0.215738  0.316837            0            0            0   
759   0.512898 -0.869396 -0.623492            0            0            0   
1253  0.199074 -1.231107 -0.516026            0            0            0   
99    0.303682  2.024294  0.316837            0            0            0   
1899  0.199074 -1.231107 -0.408560            0            0            0   
2253  0.826723 -0.507685 -0.247360            0            0            0   
3302  0.408290 -0.507685  0.854168            0            0            0   
1875  0.199074 -0.869396 -0.220494            0            0            0   
360   0.826723  0.577449  0.854168            0            0            0   
3307  0.094465 -0.507685 -0.435426            0            0            0   
4170  0.094465 -1.231107  0.048172            0            0            0   
4604  0.512898  0.215738 -0.381693            0            0            0   
1344 -2.625347 -1.954530 -0.354827            0            0            0   
4183  0.303682 -1.231107 -0.516026            0            0            0   
2765  0.512898  1.662582  1.525832            0            0            0   
442   1.558980  0.215738  0.048172            0            0            0   
517   0.512898  0.215738 -0.220494            0            0            0   
1018  0.512898  0.577449  0.182504            0            0            0   
1589  0.094465  0.215738 -0.435426            0            0            0   
2308 -1.056225 -0.507685 -0.381693            0            0            0   
1727  0.408290 -0.507685 -0.704360            0            0            0   
1021  0.512898  0.577449  0.048172            0            0            0   
4201 -1.579266 -1.592818 -0.838425            0            0            0   
1461 -0.219359  0.939160 -0.354827            0            0            0   
2086 -2.102306  1.300871 -0.623492            0            0            0   
1145  0.722114  0.215738  0.719836            0            0            0   
...        ...       ...       ...          ...          ...          ...   
2478  0.512898 -1.231107  0.075038            0            0            0   
3478  0.303682 -0.869396 -0.193627            0            0            0   
49   -0.533184 -1.231107 -0.892158            0            0            0   
2591  0.512898  1.662582  1.122834            0            0            0   
31    0.303682 -1.231107 -0.757825            0            0            0   
2303 -0.533184 -0.507685 -0.623492            0            0            0   
2769 -0.637792  1.662582 -0.139894            0            0            0   
753  -2.625347 -0.869396 -0.704092            0            0            0   
1617 -0.742400  0.215738 -0.301094            0            0            0   
750  -1.474657 -0.869396 -0.811558            0            0            0   
1027 -0.428576  0.577449 -0.354827            0            0            0   
1028 -0.219359  0.577449  0.666102            0            0            0   
4545  0.199074  0.577449 -0.193627            0            0            0   
4570  0.931331  0.215738  0.451170            0            0            0   
3975 -0.114751 -0.145974  0.854168            0            0            0   
2569  0.931331 -3.039663 -0.408560            0            0            0   
4224  0.408290 -1.592818 -0.462293            0            0            0   
2096  0.303682  1.300871  0.854168            0            0            0   
4151 -1.683874 -0.869396 -0.811558            0            0            0   
2714  0.512898  0.939160  0.370570            0            0            0   
246

In [24]:
poly = preprocessing.PolynomialFeatures(degree=2, interaction_only=False, include_bias=False)
features_135_train = pd.DataFrame(poly.fit_transform(features_scaled_train), columns=poly.get_feature_names(features_scaled_train.columns))
features_135_train.head()

Alcohol    Points     Price  Year_2005.0  Year_2007.0  Year_2008.0  \
0  0.279669 -0.116308 -0.389654          0.0          0.0          0.0   
1 -0.247591 -2.267960 -0.648111          0.0          0.0          0.0   
2 -0.247591 -1.909352 -0.648111          0.0          0.0          0.0   
3  0.490573 -0.474917 -0.615804          0.0          0.0          0.0   
4  0.596025  0.959518  0.579557          0.0          0.0          0.0   

   Year_2009.0  Year_2010.0  Year_2011.0  Year_2012.0           ...            \
0          0.0          0.0          0.0          0.0           ...             
1          0.0          0.0          0.0          0.0           ...             
2          0.0          0.0          0.0          0.0           ...             
3          0.0          0.0          0.0          0.0           ...             
4          0.0          0.0          0.0          0.0           ...             

   Appellation_ZD^2  Appellation_ZD Appellation_Ziata  \
0               0.0                               0.0   
1               0.0                               0.0   
2               0.0                               0.0   
3               0.0                               0.0   
4               0.0                               0.0   

   Appellation_ZD Appellation_Zinfandelic  \
0                                     0.0   
1                                     0.0   
2                                     0.0   
3                                     0.0   
4                                     0.0   

   Appellation_ZD Appellation_Zotovich  Appellation_Ziata^2  \
0                                  0.0                  0.0   
1                                  0.0                  0.0   
2                                  0.0                  0.0   
3                                  0.0                  0.0   
4                                  0.0                  0.0   

   Appellation_Ziata Appellation_Zinfandelic  \
0                                        0.0   
1                                        0.0   
2                                        0.0   
3                                        0.0   
4                                        0.0   

   Appellation_Ziata Appellation_Zotovich  Appellation_Zinfandelic^2  \
0                                     0.0                        0.0   
1                                     0.0                        0.0   
2                                     0.0                        0.0   
3                                     0.0                        0.0   
4                                     0.0                        0.0   

   Appellation_Zinfandelic Appellation_Zotovich  Appellation_Zotovich^2  
0                                           0.0                     0.0  
1                                           0.0                     0.0  
2                                           0.0                     0.0  
3                                           0.0                     0.0  
4                                           0.0                     0.0  

[5 rows x 578349 columns]

In [ ]:
from sklearn.feature_selection import VarianceThreshold
thresholder = VarianceThreshold(threshold=.5)

def variance_threshold_selector(data, threshold=0.5):
    selector = VarianceThreshold(threshold)
    selector.fit(data)
    return data[data.columns[selector.get_support(indices=True)]]

features_selected_train = variance_threshold_selector(features_135_train)

features_selected_train.head()

In [ ]:
import seaborn as sns

sns.set(style="white")


# Compute the correlation matrix
corr = features_selected_train.corr()

# Generate a mask for the upper triangle
mask = np.zeros_like(corr, dtype=np.bool)
mask[np.triu_indices_from(mask)] = True

# Set up the matplotlib figure
f, ax = plt.subplots(figsize=(11, 9))

# Generate a custom diverging colormap
cmap = sns.diverging_palette(220, 10, as_cmap=True)

# Draw the heatmap with the mask and correct aspect ratio
sns.heatmap(corr, mask=mask, cmap=cmap, vmax=.3, center=0,
            square=True, linewidths=.5, cbar_kws={"shrink": 1})

In [ ]:
#  Create correlation matrix
corr_matrix = features_selected_train.corr().abs()

# Select upper triangle of correlation matrix
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(np.bool))

# Find index of feature columns with correlation greater than 0.95
to_drop = [column for column in upper.columns if any(upper[column] > 0.95)]
def calculate_cols_to_drop(upper):
    #to_drop = [column for column in upper.columns if any(upper[column] > 0.95)]
    shape = upper.shape
    print(shape)
    row_index = upper.index
    col_index = upper.columns
    to_drop = []
    
    for row in range(0, shape[0]):
        for col in range(0, shape[1]):
            #print(upper.iloc[row,col])
            if upper.iloc[row, col] > 0.95:
                if row_index[row] in to_drop or col_index[col] in to_drop:
                    pass
                else:
                    to_drop.append(col_index[col])
    
    temp = upper.drop(index= to_drop, columns=to_drop)
    
    print(len(to_drop))
    
    shape = temp.shape
    for row in range(0, shape[0]):
        for col in range(0, shape[1]):
            if temp.iloc[row, col] > 0.95:
                print("It did not work")
                print('row: ' + temp.index[row])
                print('col: ' + temp.columns[col])
            
    return to_drop

corr_removed = calculate_cols_to_drop(upper)

print(len([column for column in upper.columns if any(upper[column] > 0.95)]))
corr_removed


In [ ]:
upper